protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+    - Once you start working on the breaking change, bump the version to `3.0.0-beta.1`.
+    - Continue to bump the beta version as you make changes. When the contract is ready for release, bump the version to `3.0.0-rc.1`.
+- New contracts start at `0.y.z` and only become `1.0.0` when they are ready for production.
+
+Versioning for monorepo releases:
+
+- Monorepo releases continue to follow the `op-contracts/vX.Y.Z` naming convention.
+- The version used for the next release is determined by the highest version bump of any individual contract in the release.
+  - Example 1: The monorepo is at `op-contracts/v1.5.0`. Various tech debt is cleaned up in contracts, but no features are added, so all contracts only bump the patch version. The next monorepo release will be `op-contracts/v1.5.1`.
+  - Example 2: The monorepo is at `op-contracts/v1.5.1`. Legacy `ALL_CAPS()` getter methods are removed from a contract, causing that contract to bump the major version. The next monorepo release will be `op-contracts/v2.0.0`.
+- A monorepo contracts release must map to an exact set of contract semvers, and this mapping must be defined in the contract release notes which are the source of truth. See [`op-contracts/v1.4.0-rc.4`](https://github.com/ethereum-optimism/optimism/releases/tag/op-contracts%2Fv1.4.0-rc.4) for an example of what release notes should look like.
+
+The above gets us most of the way there, and there is one problem not yet addressed: When a release is proposed to governance, we propose a commit hash, and often those proposed contracts are already deployed to mainnet.
+For example, the [Fault Proofs governance proposal](https://gov.optimism.io/t/upgrade-proposal-fault-proofs/8161) provides specific addresses that will be used.
+This is ideal and we should continue to do that, but we need to fit this in to the above scheme.
+
+To accommodate this, once ready for governance approval the release flow is:
+
+- On the `develop` branch, bump the version of all contracts to their respective `X.Y.Z-rc.n`. The `X.Y.Z` here refers to the contract-specific versions so differs per-contract, whereas the `.rc.n` is the same across all contracts.
+- Branch off of `develop` and create a branch name `proposal/op-contracts/vX.Y.Z`. Here, `X.Y.Z` is the new version of the monorepo release.
+- On this branch, create a new commit that removes the `-rc.n` from all contracts. This is the commit hash that will be used to deploy the contracts and proposed to governance.
+- Once governance approves the proposal, merge the proposal branch into `develop` and bump the version of all contracts to their respective `X.Y.Z`. When there are conflicts, resolve them in favor of the `develop` branch. In other words:
+  - If a contract has not had any changes since the proposal, the version should be bumped to `X.Y.Z`.
+  - If a contract has had changes since the proposal, the version should be stay as whatever is the latest version on `develop`.
```

Comment:
I'm having trouble understanding this section. Specifically how to choose the `X.Y.Z`.

Reply:
I will try to clarify with examples, and can move them into the doc if helpful:

- ContractA is `1.2.3-rc1` on develop.
- We create the release branch, and the final ContractA version on that branch is `1.2.3`
- Governance approves the release so we want to merge the final versions of the contracts, including ContractA, back into develop.

There are two scenarios for this PR that merges the release branch into develop:

1. On develop, no changes have been made to ContractA. The PR therefore changes ContractA's version on develop from `1.2.3-rc.1` to `1.2.3`, and no other changes to ContractA occur.
2. On develop, breaking changes have been made to ContractA for a new feature, and it's currently versioned as `2.0.0-beta.3`. The PR should modify the source code of ContractA appropriately and bump the version to `2.0.0-beta.4`